src/qt/walletmodel.cpp

```diff
@@ -854,18 +854,39 @@ bool WalletModel::isSpent(const COutPoint& outpoint) const
 }
 
 // AvailableCoins + LockedCoins grouped by wallet address (put change in one group with wallet address)
-void WalletModel::listCoins(std::map<QString, std::vector<COutput> >& mapCoins) const
+void WalletModel::listCoins(std::map<ListCoinsKey, std::vector<COutput> >& mapCoins) const
 {
     CWallet::AvailableCoinsFilter filter;
     filter.fIncludeLocked = true;
     std::vector<COutput> vCoins;
     wallet->AvailableCoins(&vCoins, nullptr, filter);
 
     for (const COutput& out : vCoins) {
-        CTxDestination address;
-        if (!out.fSpendable || !ExtractDestination(out.tx->vout[out.i].scriptPubKey, address))
-            continue;
-        mapCoins[QString::fromStdString(EncodeDestination(address))].emplace_back(out);
+        if (!out.fSpendable) continue;
+
+        const CScript& scriptPubKey = out.tx->vout[out.i].scriptPubKey;
+        const bool isP2CS = scriptPubKey.IsPayToColdStaking();
+
+        CTxDestination outputAddress;
+        CTxDestination outputAddressStaker;
+        if (isP2CS) {
+            txnouttype type; std::vector<CTxDestination> addresses; int nRequired;
+            if(!ExtractDestinations(scriptPubKey, type, addresses, nRequired)
+                        && addresses.size() == 2) continue;
```

Comment:
I think this should be
```suggestion
            if(!ExtractDestinations(scriptPubKey, type, addresses, nRequired)
                        || addresses.size() != 2) continue;
```
and we could probably also log an error (actually we could even throw, as this should  never happen).

Reply:
yeah, done.